### PR TITLE
Make use of TimeZoneInfo.CachedData on Linux.

### PR DESCRIPTION
**Pull request by eerhardt**
I refactored the TimeZoneInfo code to share more code between Windows and Linux.  This allows us to use the CachedData object effectively on both platforms.

I also fixed a small typo in a method name.

Fix https://github.com/dotnet/corefx/issues/2497.

### CODE REVIEW

**Comment by eerhardt** [comment]
@tarekgh, @ellismg, @joshfree

File: src/mscorlib/src/System/TimeZoneInfo.cs [hunk 1]
```diff
@@ -2154,64 +2149,40 @@ static public TimeZoneInfo FindSystemTimeZoneById(string id)
...
             }
-            catch (UnauthorizedAccessException e)
-            {
-                throw new SecurityException(Environment.GetResourceString("Security_CannotReadFileData", id), e);
-            }
-            catch (FileNotFoundException e)
+            catch (UnauthorizedAccessException ex)
             {
-                throw new TimeZoneNotFoundException(Environment.GetResourceString("TimeZoneNotFound_MissingData", id), e);
+                value = null;
```
Comment (tarekgh) [inline src/mscorlib/src/System/TimeZoneInfo.cs]: can you just initialize the value with null in the beginning and remove it from the rest.

**Comment by tarekgh** [comment]
LGTM

File: src/mscorlib/src/System/TimeZoneInfo.cs [hunk 2]
```diff
@@ -1005,8 +982,42 @@ private AdjustmentRule GetPreviousAdjustmentRule(AdjustmentRule rule)
...
+
+                if (reg != null) {
+                    foreach (string keyName in reg.GetSubKeyNames()) {
+                        TimeZoneInfo value;
+                        Exception ex;
+                        TryGetTimeZone(keyName, false, out value, out ex, cachedData);  // populate the cache
+                    }
+                }
+            }
+#else // FEATURE_WIN32_REGISTRY
```
Comment (stephentoub) [inline src/mscorlib/src/System/TimeZoneInfo.cs]: Should this be:
```C#
#elif PLATFORM_UNIX
```
to keep it consistent with TryGetTimeZoneByFile being defined uned PLATFORM_UNIX?

Comment (eerhardt) [inline src/mscorlib/src/System/TimeZoneInfo.cs]: The conditionals in this file are a little messy, see https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/TimeZoneInfo.cs#L2071.

I can change both of these locations as requested.  But that means if someone tries compiling mscorlib without one of these conditionals set to true, the compile is going to fail.  I'm assuming that is exactly what we want... correct?

File: src/mscorlib/src/System/TimeZoneInfo.cs (hunk 1 above)
Comment (eerhardt) [inline src/mscorlib/src/System/TimeZoneInfo.cs]: Good idea.  Done.

File: src/mscorlib/src/System/TimeZoneInfo.cs (hunk 2 above)
Comment (stephentoub) [inline src/mscorlib/src/System/TimeZoneInfo.cs]: I think so.  Thanks.

**Comment by eerhardt** [comment]
@ellismg,

> You could consider factoring this into a method called IsValidSystemTimeZoneId() 

Good idea.  Will do.

> This dance feels really weird here. Would it make more sense for TryGetTimeZoneByFile to just ensure that the exception type is always an InvalidTimeZoneException when result == TimeZoneInfoResult.InvalidTimeZoneException?

I totally agree it felt weird writing it.  Normally in a situation like this, I would add `Debug.Assert(e is InvalidTimeZoneException)`, but there doesn't appear to be a Debug class in System.Diagnostics.  Is BCLDebug.Assert the right thing to do here?

**Comment by ellismg** [comment]
@eerhardt, I think Contract.Assert is preferable.  BCLDebug predates it, but it looks like Contract.Assert is used more often.  We should probably look into rationalizing these two at some point.

**Comment by eerhardt** [comment]
I've responded to all feedback.  Please let me know if there is any more.

**Comment by akoeplinger** [comment]
@ellismg I was under the impression that Contract.Assert is discouraged: https://github.com/dotnet/corefx/pull/1503

**Pull request merged**